Stack/Opc.Ua.Core/Types/Utils/AsyncAutoResetEvent.cs

```diff
@@ -0,0 +1,105 @@
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ * 
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * The complete license agreement can be found here:
+ * http://opcfoundation.org/License/MIT/1.00/
+ * ======================================================================*/
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Opc.Ua.Types.Utils
+{
+    /// <summary>
+    /// An async version of <see cref="AutoResetEvent"/> based on
+    /// https://devblogs.microsoft.com/pfxteam/building-async-coordination-primitives-part-2-asyncautoresetevent/.
+    /// </summary>
+    public class AsyncAutoResetEvent
+    {
+        private static readonly Task s_completed = Task.FromResult(true);
+        private readonly Queue<TaskCompletionSource<bool>> m_waits = new Queue<TaskCompletionSource<bool>>();
+        private bool m_signaled;
+
+        /// <summary>
+        /// Task can wait for next event.
+        /// </summary>
+        public Task WaitAsync()
+        {
+            lock (m_waits)
+            {
+                if (m_signaled)
+                {
+                    m_signaled = false;
+                    return s_completed;
+                }
+                else
+                {
+                    var tcs = new TaskCompletionSource<bool>();
+                    m_waits.Enqueue(tcs);
+                    return tcs.Task;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Set next waiting task.
+        /// </summary>
+        public void Set()
+        {
+            TaskCompletionSource<bool> toRelease = null;
+            lock (m_waits)
+            {
+                if (m_waits.Count > 0)
+                {
+                    toRelease = m_waits.Dequeue();
+                }
+                else
+                {
+                    m_signaled = true;
+                    return;
+                }
+            }
+            toRelease?.SetResult(true);
+        }
+
+        /// <summary>
+        /// Set all waiting tasks.
+        /// </summary>
+        public void SetAll()
+        {
+            TaskCompletionSource<bool> toRelease = null;
+            lock (m_waits)
+            {
+                while (m_waits.Count > 0)
+                {
+                    toRelease = m_waits.Dequeue();
```

Comment:
I'm guessing it's another style guide matter, but why is `toRelease` not declared closer to its point of usage within the loop? Same goes for `Set()`. You've used `var` nearby to the actual usage in `WaitAsync` in a way that make perfect sense and avoids leaking variables out of their relevant context.

The code still looks correct to me so there's no problem, but it did pique my interest

Reply:
ok